File: tome-backend/tome-user-data/src/main/java/com/safegergis/tome_user_data/service/UserBookService.java
package com.safegergis.tome_user_data.service;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.safegergis.tome_user_data.dto.BookSummaryDTO;
import com.safegergis.tome_user_data.dto.CreateUserBookRequest;
import com.safegergis.tome_user_data.dto.UpdateUserBookRequest;
import com.safegergis.tome_user_data.dto.UserBookDTO;
import com.safegergis.tome_user_data.enums.ReadingStatus;
import com.safegergis.tome_user_data.exception.DuplicateResourceException;
import com.safegergis.tome_user_data.exception.ForbiddenException;
import com.safegergis.tome_user_data.exception.ResourceNotFoundException;
import com.safegergis.tome_user_data.mapper.UserBookMapper;
import com.safegergis.tome_user_data.model.UserBook;
import com.safegergis.tome_user_data.repository.UserBookRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

/**
 * Service layer for UserBook operations
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class UserBookService {

    private final UserBookRepository userBookRepository;
    private final BookServiceClient bookServiceClient;

    /**
     * Add a book to user's shelf
     */
    @Transactional
    public UserBookDTO addBookToShelf(Long userId, CreateUserBookRequest request) {
        log.debug("Adding book {} to shelf for user {}", request.getBookId(), userId);

        // Check if book exists in tome-content
        BookSummaryDTO book = bookServiceClient.getBook(request.getBookId());

        // Check for duplicate
        if (userBookRepository.existsByUserIdAndBookId(userId, request.getBookId())) {
            throw new DuplicateResourceException("UserBook",
                    "user " + userId + " and book " + request.getBookId());
        }

        // Create user book
        UserBook userBook = UserBook.builder()
                .userId(userId)
                .bookId(request.getBookId())
                .status(request.getStatus())
                .currentPage(request.getCurrentPage() != null ? request.getCurrentPage() : 0)
                .currentSeconds(request.getCurrentSeconds() != null ? request.getCurrentSeconds() : 0)
                .userPageCount(request.getUserPageCount())
                .userAudioLengthSeconds(request.getUserAudioLengthSeconds())
                .personalRating(request.getPersonalRating())
                .notes(request.getNotes())
                .build();

        // Set timestamps based on status
        if (request.getStatus() == ReadingStatus.CURRENTLY_READING) {
            userBook.setStartedAt(OffsetDateTime.now());
        } else if (request.getStatus() == ReadingStatus.READ) {
            userBook.setFinishedAt(OffsetDateTime.now());
        }

        UserBook saved = userBookRepository.save(userBook);
        return UserBookMapper.toDTO(saved, book);
    }

    /**
     * Get a specific user book
     */
    @Transactional(readOnly = true)
    public UserBookDTO getUserBook(Long userId, Long userBookId) {
        UserBook userBook = userBookRepository.findById(userBookId)
                .orElseThrow(() -> new ResourceNotFoundException("UserBook", userBookId));

        // Check ownership
        if (!userBook.getUserId().equals(userId)) {
            throw new ForbiddenException("You do not have permission to access this resource");
        }

        BookSummaryDTO book = bookServiceClient.getBook(userBook.getBookId());
        return UserBookMapper.toDTO(userBook, book);
    }

    /**
     * Get all books for a user, optionally filtered by status
     * This is the key method requested by the user
     */
    @Transactional(readOnly = true)
    public List<UserBookDTO> getUserBooksByStatus(Long userId, ReadingStatus status) {
        log.debug("Fetching books for user {} with status {}", userId, status);

        List<UserBook> userBooks;

        if (status == null) {
            // Get all books
            userBooks = userBookRepository.findByUserId(userId);
        } else {
            // Filter by status - convert enum to string for native query
            userBooks = userBookRepository.findByUserIdAndStatus(userId, status.name());
        }

        // Enrich with book details and map to DTO
        return userBooks.stream()
                .map(userBook -> {
                    BookSummaryDTO book = bookServiceClient.getBook(userBook.getBookId());
                    return UserBookMapper.toDTO(userBook, book);
                })
                .collect(Collectors.toList());
    }

    /**
     * Update a user book
     */
    @Transactional
    public UserBookDTO updateUserBook(Long userId, Long userBookId, UpdateUserBookRequest request) {
        UserBook userBook = userBookRepository.findById(userBookId)
                .orElseThrow(() -> new ResourceNotFoundException("UserBook", userBookId));

        // Check ownership
        if (!userBook.getUserId().equals(userId)) {
            throw new ForbiddenException();
        }

        // Update fields if provided
        if (request.getStatus() != null) {
            updateStatus(userBook, request.getStatus());
        }
        if (request.getCurrentPage() != null) {
            userBook.setCurrentPage(request.getCurrentPage());
        }
        if (request.getCurrentSeconds() != null) {
            userBook.setCurrentSeconds(request.getCurrentSeconds());
        }
        if (request.getUserPageCount() != null) {
            userBook.setUserPageCount(request.getUserPageCount());
        }
        if (request.getUserAudioLengthSeconds() != null) {
            userBook.setUserAudioLengthSeconds(request.getUserAudioLengthSeconds());
        }
        if (request.getPersonalRating() != null) {
            userBook.setPersonalRating(request.getPersonalRating());
        }
        if (request.getNotes() != null) {
            userBook.setNotes(request.getNotes());
        }

        UserBook updated = userBookRepository.save(userBook);
        BookSummaryDTO book = bookServiceClient.getBook(updated.getBookId());
        return UserBookMapper.toDTO(updated, book);
    }

    /**
     * Update reading status with appropriate timestamp handling
     */
    @Transactional
    public UserBookDTO updateReadingStatus(Long userId, Long userBookId, ReadingStatus newStatus) {
        UserBook userBook = userBookRepository.findById(userBookId)
                .orElseThrow(() -> new ResourceNotFoundException("UserBook", userBookId));

        // Check ownership
        if (!userBook.getUserId().equals(userId)) {
            throw new ForbiddenException();
        }

        updateStatus(userBook, newStatus);

        UserBook updated = userBookRepository.save(userBook);
        BookSummaryDTO book = bookServiceClient.getBook(updated.getBookId());
        return UserBookMapper.toDTO(updated, book);
    }

    /**
     * Helper method to update status and set appropriate timestamps
     */
    private void updateStatus(UserBook userBook, ReadingStatus newStatus) {
        ReadingStatus oldStatus = userBook.getStatus();
        userBook.setStatus(newStatus);

        // Set started_at when moving to currently-reading
        if (newStatus == ReadingStatus.CURRENTLY_READING && userBook.getStartedAt() == null) {
            userBook.setStartedAt(OffsetDateTime.now());
        }

        // Set finished_at when moving to read
        if (newStatus == ReadingStatus.READ && oldStatus != ReadingStatus.READ) {
            userBook.setFinishedAt(OffsetDateTime.now());
        }

        // Set dnf_date when marking as did-not-finish
        if (newStatus == ReadingStatus.DID_NOT_FINISH && oldStatus != ReadingStatus.DID_NOT_FINISH) {
            userBook.setDnfDate(OffsetDateTime.now());
        }
    }

    /**
     * Mark a book as Did Not Finish
     */
    @Transactional
    public UserBookDTO markAsDidNotFinish(Long userId, Long userBookId, String reason) {
        UserBook userBook = userBookRepository.findById(userBookId)
                .orElseThrow(() -> new ResourceNotFoundException("UserBook", userBookId));

        // Check ownership
        if (!userBook.getUserId().equals(userId)) {
            throw new ForbiddenException();
        }

        userBook.setStatus(ReadingStatus.DID_NOT_FINISH);
        userBook.setDnfDate(OffsetDateTime.now());
        userBook.setDnfReason(reason);

        UserBook updated = userBookRepository.save(userBook);
        BookSummaryDTO book = bookServiceClient.getBook(updated.getBookId());
        return UserBookMapper.toDTO(updated, book);
    }

    /**
     * Remove a book from user's shelf
     */
    @Transactional
    public void removeBookFromShelf(Long userId, Long userBookId) {
        UserBook userBook = userBookRepository.findById(userBookId)
                .orElseThrow(() -> new ResourceNotFoundException("UserBook", userBookId));

        // Check ownership
        if (!userBook.getUserId().equals(userId)) {
            throw new ForbiddenException();
        }

        userBookRepository.delete(userBook);
    }
}

File: tome-frontend/tome/services/user-book.service.ts
/**
 * User Book API Service
 * Handles communication with tome-user-data service for UserBook operations
 */

import { ENV } from '@/config/env';
import { UserBookDTO } from '@/types/reading-session';
import { apiClient } from './api-client';

/**
 * Request payload for updating a user book
 */
export interface UpdateUserBookRequest {
    status?: 'WANT_TO_READ' | 'CURRENTLY_READING' | 'READ' | 'DID_NOT_FINISH';
    currentPage?: number;
    currentSeconds?: number;
    userPageCount?: number;
    userAudioLengthSeconds?: number;
    personalRating?: number;
    notes?: string;
}

/**
 * Request payload for adding a book to shelf
 */
export interface CreateUserBookRequest {
    bookId: number;
    status: 'WANT_TO_READ' | 'CURRENTLY_READING' | 'READ' | 'DID_NOT_FINISH';
}

/**
 * User Book API endpoints
 */
export const userBookApi = {
    /**
     * Get user books by status
     * @param token - JWT authentication token
     * @param status - Optional reading status filter (currently-reading, want-to-read, read, did-not-finish)
     * @param userId - Optional user ID to fetch another user's books (only currently-reading supported for privacy)
     * @returns List of user books
     */
    getUserBooks: async (
        token: string,
        status?: 'currently-reading' | 'want-to-read' | 'read' | 'did-not-finish',
        userId?: number
    ): Promise<UserBookDTO[]> => {
        console.log('[userBookApi] Fetching user books:', { status, userId });

        let url = `${ENV.USER_DATA_API_URL}/user-books`;
        const params = new URLSearchParams();

        if (status) {
            params.append('status', status);
        }
        if (userId) {
            params.append('userId', userId.toString());
        }

        if (params.toString()) {
            url += `?${params.toString()}`;
        }

        const result = await apiClient.authenticatedFetch<UserBookDTO[]>(url, token);

        console.log('[userBookApi] User books fetched successfully:', result.length);
        return result;
    },

    /**
     * Add a book to user's shelf
     * @param bookId - The Book ID
     * @param status - Initial reading status
     * @param token - JWT authentication token
     * @returns Created user book
     */
    addBookToShelf: async (
        bookId: number,
        status: 'WANT_TO_READ' | 'CURRENTLY_READING' | 'READ' | 'DID_NOT_FINISH',
        token: string
    ): Promise<UserBookDTO> => {
        console.log('[userBookApi] Adding book to shelf:', { bookId, status });

        const result = await apiClient.authenticatedFetch<UserBookDTO>(
            `${ENV.USER_DATA_API_URL}/user-books`,
            token,
            {
                method: 'POST',
                body: JSON.stringify({ bookId, status }),
            }
        );

        console.log('[userBookApi] Book added to shelf successfully');
        return result;
    },

    /**
     * Update reading status of a user book
     * @param userBookId - The UserBook ID
     * @param status - New reading status
     * @param token - JWT authentication token
     * @returns Updated user book
     */
    updateReadingStatus: async (
        userBookId: number,
        status: 'want-to-read' | 'currently-reading' | 'read' | 'did-not-finish',
        token: string
    ): Promise<UserBookDTO> => {
        console.log('[userBookApi] Updating reading status:', { userBookId, status });

        const result = await apiClient.authenticatedFetch<UserBookDTO>(
            `${ENV.USER_DATA_API_URL}/user-books/${userBookId}/status?status=${status}`,
            token,
            {
                method: 'PATCH',
            }
        );

        console.log('[userBookApi] Reading status updated successfully');
        return result;
    },

    /**
     * Mark a book as Did Not Finish
     * @param userBookId - The UserBook ID
     * @param reason - Optional reason for DNF
     * @param token - JWT authentication token
     * @returns Updated user book
     */
    markAsDidNotFinish: async (
        userBookId: number,
        token: string,
        reason?: string
    ): Promise<UserBookDTO> => {
        console.log('[userBookApi] Marking book as DNF:', { userBookId, reason });

        const url = reason
            ? `${ENV.USER_DATA_API_URL}/user-books/${userBookId}/dnf?reason=${encodeURIComponent(reason)}`
            : `${ENV.USER_DATA_API_URL}/user-books/${userBookId}/dnf`;

        const result = await apiClient.authenticatedFetch<UserBookDTO>(url, token, {
            method: 'POST',
        });

        console.log('[userBookApi] Book marked as DNF successfully');
        return result;
    },

    /**
     * Update a user book (including user overrides for page count/audio length)
     * @param userBookId - The UserBook ID
     * @param data - Fields to update
     * @param token - JWT authentication token
     * @returns Updated user book
     */
    updateUserBook: async (
        userBookId: number,
        data: UpdateUserBookRequest,
        token: string
    ): Promise<UserBookDTO> => {
        console.log('[userBookApi] Updating user book:', {
            userBookId,
            fields: Object.keys(data),
        });

        const result = await apiClient.authenticatedFetch<UserBookDTO>(
            `${ENV.USER_DATA_API_URL}/user-books/${userBookId}`,
            token,
            {
                method: 'PUT',
                body: JSON.stringify(data),
            }
        );

        console.log('[userBookApi] User book updated successfully');
        return result;
    },
};
